primitives/src/circuit/merkle_tree/sparse_merkle_tree.rs

```diff
@@ -0,0 +1,688 @@
+                    children[1].value(),
+                    branch == &0,
+                    branch == &2,
+                )),
+                _ => None,
+            })
+            .collect();
+
+        Self::new(&nodes)
+    }
+}
+
+trait SparseMerkleTreeHelperGadget<F: RescueParameter> {
+    /// Computes the merkle root based on an empty leaf element and a
+    /// merkle path.
+    /// * `path_vars` - variables corresponding to the Merkle path.
+    /// * `return` - variable corresponding to the root value of the Merkle
+    ///   tree.
+    fn compute_merkle_root_for_empty(
+        &mut self,
+        path_vars: &Rescue3AryMembershipProofVar,
+    ) -> Result<Variable, CircuitError>;
+}
+
+impl<F: RescueParameter> SparseMerkleTreeHelperGadget<F> for PlonkCircuit<F> {
+    fn compute_merkle_root_for_empty(
+        &mut self,
+        path_vars: &Rescue3AryMembershipProofVar,
+    ) -> Result<Variable, CircuitError> {
+        let mut cur_label = self.zero();
+        for cur_node in path_vars.nodes.iter() {
+            let input_labels =
+                <PlonkCircuit<F> as MerkleTreeHelperGadget<SparseMerkleTree<F>>>::permute(
+                    self,
+                    cur_label,
+                    cur_node.sibling1,
+                    cur_node.sibling2,
+                    cur_node.is_left_child,
+                    cur_node.is_right_child,
+                )?;
+            // check that the left child's label is non-zero
+            self.non_zero_gate(input_labels[0])?;
+            cur_label =
+                RescueNativeGadget::<F>::rescue_sponge_no_padding(self, &input_labels, 1)?[0];
+        }
+        Ok(cur_label)
+    }
+}
+
+impl<F: RescueParameter> MerkleTreeHelperGadget<SparseMerkleTree<F>> for PlonkCircuit<F> {
+    type MembershipProofBooleanEncoding = MembershipProofBooleanEncoding<SparseMerkleTree<F>>;
+
+    type MembershipProofVar = Rescue3AryMembershipProofVar;
+
+    fn permute(
+        &mut self,
+        node: Variable,
+        sib1: Variable,
+        sib2: Variable,
+        node_is_left: BoolVar,
+        node_is_right: BoolVar,
+    ) -> Result<[Variable; 3], CircuitError> {
+        let one = F::one();
+        let left_node = self.conditional_select(node_is_left, sib1, node)?;
+        let right_node = self.conditional_select(node_is_right, sib2, node)?;
+        let left_plus_right = self.add(left_node, right_node)?;
+        let mid_node = self.lc(
+            &[node, sib1, sib2, left_plus_right],
+            &[one, one, one, one.neg()],
+        )?;
+        Ok([left_node, mid_node, right_node])
+    }
+
+    fn constrain_membership_proof(
+        &mut self,
+        merkle_path: &MembershipProofBooleanEncoding<SparseMerkleTree<F>>,
+        pos: Index<F>,
+    ) -> Result<Rescue3AryMembershipProofVar, CircuitError> {
+        // Setup node variables
+        let nodes = merkle_path
+            .nodes
+            .clone()
+            .into_iter()
```

Comment:
Is it different from `Merkle_path.nodes.iter().map..`? Do we have to move `node` here?

Reply:
Good catch, fixed here: https://github.com/EspressoSystems/jellyfish/pull/172/commits/82ca71a9da9de693812963a3892d81613eef25d8